File: engineering/concurrency/06-dining-philosophers/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

const numPhilosophers = 5

// Chopstick 代表一根筷子，用互斥锁模拟。
type Chopstick struct {
	sync.Mutex
}

// Philosopher 代表一位哲学家。
type Philosopher struct {
	id          int
	left, right *Chopstick
}

func (p *Philosopher) Dine(wg *sync.WaitGroup, eatCount *int, mu *sync.Mutex) {
	defer wg.Done()
	for i := 0; i < 10; i++ {
		// 思考
		time.Sleep(time.Millisecond)

		// TODO: 实现就餐逻辑，避免死锁
		// 策略选项（选其一）：
		// 1. 资源排序：总是先拿编号小的筷子
		// 2. 限制同时就餐人数：最多 4 人同时尝试拿筷子
		// 3. 非对称策略：偶数 id 先左后右，奇数 id 先右后左

		p.left.Lock()
		p.right.Lock()
		// 就餐
		mu.Lock()
		*eatCount++
		mu.Unlock()
		time.Sleep(time.Millisecond)
		p.right.Unlock()
		p.left.Unlock()
	}
}

func main() {
	chopsticks := make([]*Chopstick, numPhilosophers)
	for i := 0; i < numPhilosophers; i++ {
		chopsticks[i] = &Chopstick{}
	}

	philosophers := make([]*Philosopher, numPhilosophers)
	for i := 0; i < numPhilosophers; i++ {
		philosophers[i] = &Philosopher{
			id:    i,
			left:  chopsticks[i],
			right: chopsticks[(i+1)%numPhilosophers],
		}
	}

	var wg sync.WaitGroup
	var mu sync.Mutex
	eatCount := 0

	done := make(chan struct{})
	go func() {
		for _, p := range philosophers {
			wg.Add(1)
			go p.Dine(&wg, &eatCount, &mu)
		}
		wg.Wait()
		close(done)
	}()

	select {
	case <-done:
		expected := numPhilosophers * 10
		if eatCount == expected {
			fmt.Printf("PASS: all %d meals completed without deadlock\n", eatCount)
		} else {
			fmt.Printf("FAIL: eatCount=%d (expected %d)\n", eatCount, expected)
		}
	case <-time.After(5 * time.Second):
		fmt.Println("FAIL: timeout, likely deadlock")
	}
}

File: engineering/concurrency/06-dining-philosophers/skeleton/python/main.py
# 并发模型：threading
import threading
import time

NUM_PHILOSOPHERS = 5

class Chopstick:
    def __init__(self):
        self._lock = threading.Lock()

    def acquire(self):
        self._lock.acquire()

    def release(self):
        self._lock.release()

class Philosopher:
    def __init__(self, pid, left, right):
        self.id = pid
        self.left = left
        self.right = right

    def dine(self, eat_count, mu):
        for _ in range(10):
            time.sleep(0.001)
            # TODO: 实现就餐逻辑，避免死锁
            self.left.acquire()
            self.right.acquire()
            with mu:
                eat_count[0] += 1
            time.sleep(0.001)
            self.right.release()
            self.left.release()

def main():
    chopsticks = [Chopstick() for _ in range(NUM_PHILOSOPHERS)]
    philosophers = [
        Philosopher(i, chopsticks[i], chopsticks[(i + 1) % NUM_PHILOSOPHERS])
        for i in range(NUM_PHILOSOPHERS)
    ]

    eat_count = [0]
    mu = threading.Lock()
    threads = []

    for p in philosophers:
        t = threading.Thread(target=p.dine, args=(eat_count, mu))
        threads.append(t)
        t.start()

    for t in threads:
        t.join(timeout=5)

    still_alive = any(t.is_alive() for t in threads)
    expected = NUM_PHILOSOPHERS * 10
    if not still_alive and eat_count[0] == expected:
        print(f"PASS: all {eat_count[0]} meals completed without deadlock")
    elif still_alive:
        print("FAIL: timeout, likely deadlock")
    else:
        print(f"FAIL: eat_count={eat_count[0]} (expected {expected})")

if __name__ == "__main__":
    main()
